src/permissions/Caveat.ts

```diff
@@ -120,7 +121,7 @@ export type CaveatSpecificationConstraint = {
    * The decorator function used to apply the caveat to restricted method
    * requests.
    */
-  decorator: CaveatDecorator<any>;
+  decorator?: CaveatDecorator<any>;
```

Comment:
We probably shouldn't have this property on the base `CaveatSpecificationConstraint` and only add it to the `RestrictedMethodSpecificationConstraint` type. You might need some type guards for the implementation in order to access the `decorator` property and call it, but I think this would be the better approach.

Reply:
We shouldn't add it to `RestrictedMethodSpecificationConstraint` as that doesn't specify caveats. But perhaps `RestrictedMethodCaveatSpecificationConstraint` should exist

Reply:
Yeah, that sounds reasonable!

Reply:
Done! This needs another review.